cocotb/share/lib/embed/embed.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright cocotb contributors
+// Licensed under the Revised BSD License, see LICENSE for details.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <embed.h>
+#include <cocotb_utils.h>   // xstr, utils_dyn_open, utils_dyn_sym
+#include <gpi.h>            // gpi_event_t
+#include <cstdlib>          // getenv
+#if ! defined(__linux__) && ! defined(__APPLE__)
+#include <windows.h>        // Win32 API for loading the embed impl library
+#include <string>           // string
+#endif
+
+#ifndef EMBED_IMPL_LIB
+#error "Name of embed implementation library required"
+#else
+#define EMBED_IMPL_LIB_STR xstr(EMBED_IMPL_LIB)
+#endif
+
+
+static void (*_embed_init_python)(void);
+static void (*_embed_sim_cleanup)(void);
+static int (*_embed_sim_init)(int argc, char const * const * argv);
+static void (*_embed_sim_event)(gpi_event_t level, const char *msg);
+
+static bool init_failed = false;
+
+
+#define PRINT_ERROR_WINDOZE(log_fmt) {\
+    LPSTR msg_ptr;\
+    if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER, NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_SYS_DEFAULT), reinterpret_cast<LPSTR>(&msg_ptr), 255, NULL)) {\
+        LOG_ERROR(log_fmt ": %s", msg_ptr);\
+        LocalFree(msg_ptr);\
+    } else {\
+        LOG_ERROR(log_fmt);\
+    }}
+
+
+static void * load_embed_impl_lib()
+{
+#if ! defined(__linux__) && ! defined(__APPLE__)
+    HMODULE hm = NULL;
+
+    // get a handle to the module
+    if (!GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, reinterpret_cast<LPCSTR>(&embed_init_python), &hm))
+    {
+        PRINT_ERROR_WINDOZE("Unable to open embed library");
+    }
+
+    // get the path to the module
+    char path[MAX_PATH];
+    if (!GetModuleFileNameA(hm, path, sizeof(path)))
+    {
+        PRINT_ERROR_WINDOZE("Unable to get filepath of embed library");
+    }
+
+    auto embed_lib_filepath = std::string(path);
+
+    // strip filename from path
+    auto embed_lib_dirpath = embed_lib_filepath.substr(0, embed_lib_filepath.rfind("\\"));
+
+    // combine the directory from the last step and the supplied filename
+    auto embed_impl_lib_path = embed_lib_dirpath + "\\" + EMBED_IMPL_LIB_STR;
+
+    return utils_dyn_open(embed_impl_lib_path.c_str());
+#else
+    return utils_dyn_open(EMBED_IMPL_LIB_STR);
+#endif
+}
+
+
+extern "C" void embed_init_python(void)
+{
+    // preload python library
+    auto libpython_path = getenv("LIBPYTHON_LOC");
```

Comment:
Maybe can store the path to libpython during setup and then load from c? No need for an extra environmental variables?
Use something like here: https://stackoverflow.com/questions/43409167/find-location-of-loaded-shared-library-from-in-that-shared-library to get the path?

Reply:
I'm not sure what you are proposing?

Reply:
Not to use environmental variable but read from some file created on setup?

Reply:
If by setup you mean package installation, no. In fact that's what we are trying to avoid.
If you mean setup like right before we call the simulator then it's possible. It is roughly equivalent to using an environment variable, but involves writing and reading from disk, which seem unnecessary.

Reply:
Not during compilation but some post install script? Should work even with binary distribution.

Reply:
Maybe ...

Reply:
I looked into that and wheels don't support running arbitrary post-install code. We could have the behavior run once and cache in the site-packages, but I'm not sure if that kosher (?). Since this is an implementation detail the user doesn't need to be involved in, we can always change it later once we know. I'll ask on the PyPa discourse.

Reply:
See [this answer](https://stackoverflow.com/a/58331693/6614127) about post-install scripts.

Reply:
In any case maybe should implement `cocotb-config --libpython-loc`?

Reply:
Maybe use `std::system`? 
It would be just nice not to have to set variables in custom flows.

Reply:
We can certainly add `cocotb-config --libpython` to help out, but we can't run that from C++, since LD_LIBRARY_PATH or PATh may be modified by that point, affecting results. This must be run before the simulator is started, there isn't really a way around it.

Reply:
>LD_LIBRARY_PATH or PATh may be modified by that point, affecting results. 

This cannot influence the position of `cocotb-config`. How?

Reply:
I can imagine some crazy so `cocotb-config` will fail. Load wrong python interpreter?

Reply:
Yeah. Pick up the wrong libpython, wrong `PYTHONHOME` is deduced.

Reply:
What I can do to help is to make the `getenv` code check for `nullptr` and default to `"libpython{version}.{ext}"`, so you get standard lookup if it isn't specified. This way custom flows that trust their environment/simulator don't have to set anything extra.

Reply:
😭

Reply:
> What I can do to help is to make the `getenv` code check for `nullptr` and default to `"libpython{version}.{ext}"`, so you get standard lookup if it isn't fully specified. This way custom flows that trust their environment/simulator don't have to set anything extra.

After some thinking, maybe better to fail with proper error than some nonsense.

Reply:
Done.

Reply:
From some simple testing, it looks `cocotb-config` can be called and is working fine.